indexing_utilities.cpp

```diff
@@ -242,6 +245,66 @@ namespace irods {
                                 generate_delay_execution_parameters());
         }
 
+
+
+
+        class  path_format_error: public std::runtime_error {
+                public:
+                        path_format_error(const std::string & s) : std::runtime_error(s.c_str()) {}
+        };
+
+        template <typename T>
+        class path_calc_and_cache {
+        public:
+            using string = std::string;
+            using idx_infos = std::set<T>;
+        private:
+            std::map<string, idx_infos> m;
+            using calc_function = std::function<idx_infos (const string& s)> ;
+            calc_function calc_;
+        public:
+            path_calc_and_cache(calc_function f): calc_{f}
+            {
+            }
+
+            idx_infos accum (const string &path)
+            {
+                auto pos = path.find_last_of("/");
+                if (path.size() < 1 || pos == string::npos)
+                    throw path_format_error{ "Couldn't parse: "s + path };
+                if (path == "/")
+                    return calc("/");
+                auto parent_len = (pos == 0) ? 1 :pos;
+                auto st = accum(path.substr(0,parent_len));
+                auto st_i = calc(path);
+                st.insert (st_i.begin(), st_i.end());
+                return st;
+            }
+
+            idx_infos calc(const string & p) { // Get individual contribution for last elem of path 'p'.
+                                               // Use cached value if it exists.
+                idx_infos i;
+                try{
+                    i = m.at(p);
+                }
+                catch(const std::out_of_range &e) {
+                    i = (this->calc_)(p);
+                }
+                m[p] = i;
+                return i;
+            }
+
+        };
+
+        class index_info final : public std::tuple<std::string,std::string,std::string> {
```

Comment:
I made it a struct but then it needed an `operator<` to be defined.  I think, because I place these values into a `std::set`.  I will leave it this way (a298592515e472f553955e219533504f984643d6), in interest of shorter compilation times over directness of expression, but am open to comments and could  though if anyone feels strongly though @alanking @korydraughn